src/Hamiltonians/GutzwillerSampling.jl

```diff
@@ -0,0 +1,102 @@
+"""
+    GutzwillerSampling(::AbstractHamiltonian; g)
+
+Wrapper over any `AbstractHamiltonian` that implements Gutzwiller sampling. In this
+importance sampling scheme the Hamiltonian is modified as follows.
+
+```math
+\\tilde{H}_{ij} = H_{ij} e^{-g(H_{jj} - H_{ii})}
+```
+
+# Example
+
+```jldoctest
+julia> H = HubbardMom1D(BoseFS{3}((1,1,1)); u=6.0, t=1.0)
+HubbardMom1D(BoseFS{3,3}((1, 1, 1)); u=6.0, t=1.0)
+
+julia> I = GutzwillerSampling(H, g=0.3)
+GutzwillerSampling(HubbardMom1D(BoseFS{3,3}((1, 1, 1)); u=6.0, t=1.0); g=0.3)
+
+julia> get_offdiagonal(H, BoseFS((2, 1, 0)), 1)
+(BoseFS{3,3}((1, 0, 2)), 2.0)
+
+julia> get_offdiagonal(I, BoseFS((2, 1, 0)), 1)
+(BoseFS{3,3}((1, 0, 2)), 0.8131393194811987)
+```
+"""
+struct GutzwillerSampling{A,T,H<:AbstractHamiltonian{T},G} <: AbstractHamiltonian{T}
+    hamiltonian::H
+end
```

Comment:
It would be good to document what the type parameters stand for. This is not relevant for using the functionality but would be really useful for reading the code. Can we maybe just write that as comments into the code?

Reply:
This is an internal thing that tells the Hamiltonian if it's an adjoint or not. Added a comment.